File: projections/typescript/foundation/hlp.foundation.rule-runtime/src/eval-support.ts
import { Codes } from './codes.js'
import type { Json, RuleError, ValueState } from './model.js'
import type { RuleEngineLimits } from './limits.js'

/** The resolution of a referenced cell handed to the evaluator. */
export interface RefValue {
  state: ValueState
  value?: Json
  error?: RuleError
}

export const refResolved = (value: Json | undefined): RefValue => ({ state: 'Resolved', value })
export const refError = (error: RuleError): RefValue => ({ state: 'Error', error })
export const refPending: RefValue = { state: 'Pending' }

/**
 * The ONE bad-aggregate refusal shape in this tier (ticket 162): an aggregate the evaluation
 * context cannot provide refuses with `rule.bad_reference` carrying the address params
 * `{agg: "section/fn/col"}`. Both resolvers (the form graph's CellResolver and the guard
 * evaluator's ContextBagResolver) construct it HERE, so the param order cannot drift within
 * the tier; the shared corpus pins the shape across tiers.
 */
export const unavailableAggregate = (fn: string, section: string, col: string): RefValue =>
  refError(err(Codes.badReference, 'agg', `${section}/${fn}/${col}`))

/** Resolves `var` / `agg` references for one rule evaluation (SPINE-1 §2.2). */
export interface ValueResolver {
  resolveVar(path: string): RefValue
  resolveAgg(fn: string, section: string, col: string): RefValue
}

/** The shared step budget + liveness guard for one whole-instance evaluation (SPINE-1 §4). */
export class EvalBudget {
  steps = 0
  cancelled = false
  private readonly deadline: number
  constructor(readonly limits: RuleEngineLimits, private readonly signal?: AbortSignal) {
    // Advisory client-tier wall-clock (the .NET integrity tier holds the authoritative one): wire
    // the previously-vestigial `cancelled`/`wallClockMs` to a real deadline (finding F4). The optional
    // `signal` is the TS analog of the .NET `CancellationToken` — an explicit liveness/cancel seam.
    this.deadline = limits.wallClockMs > 0 ? Date.now() + limits.wallClockMs : Number.POSITIVE_INFINITY
  }

  private checkClock(): void {
    // The wall-clock / abort signal is a NON-authoritative liveness guard (D1 ratification 2026-07-01):
    // it throws RuleTimeout, which PROPAGATES as an infrastructure fault and never becomes a divergent
    // evaluation outcome. The op-budget (`steps > stepBudget`) is the sole authoritative bound.
    if (this.cancelled || this.signal?.aborted || Date.now() >= this.deadline) { this.cancelled = true; throw new RuleTimeout() }
  }

  charge(): void {
    this.checkClock()
    if (++this.steps > this.limits.stepBudget) throw new RuleBudget()
  }

  /** Charges `size` steps proportional to a large operand (string concat / money arithmetic), so a
   *  single big operand cannot run uncharged (finding F4); fails closed identically to `charge`. */
  chargeSize(size: number): void {
    this.checkClock()
    this.steps += size > this.limits.stepBudget ? this.limits.stepBudget + 1 : (size < 1 ? 1 : size)
    if (this.steps > this.limits.stepBudget) throw new RuleBudget()
  }
}

/** Per-cell evaluation state. */
export interface EvalContext {
  resolver: ValueResolver
  now: Date
  budget: EvalBudget
}

/** A recoverable error → becomes a ComputedValue.Error / Validity failure. */
export class RuleEvalError extends Error {
  constructor(readonly error: RuleError) {
    super(error.code)
  }
}

/** A referenced cell is Pending → the whole outcome becomes Pending. */
export class RulePending extends Error {}

/** The per-instance step budget was exhausted — the authoritative, deterministic, outcome-affecting
 *  bound (becomes a fail-closed `rule.budget_exceeded` result). */
export class RuleBudget extends Error {}

/**
 * The wall-clock liveness guard / abort signal tripped. This is a NON-authoritative infrastructure
 * fault (D1 ratification 2026-07-01): it PROPAGATES out of the evaluator — it is NEVER converted into
 * an evaluation outcome, so the (time/hardware-dependent) wall-clock cannot make this tier disagree
 * with the .NET tier. The .NET analog is `RuleEngineTimeoutException`. Consumers catch it as an
 * infrastructure fault (retry / degrade), never as a rule verdict.
 */
export class RuleTimeout extends Error {}

export const err = (code: string, key?: string, value?: string): RuleError => ({
  code,
  params: key !== undefined && value !== undefined ? { [key]: value } : {},
})

File: projections/dotnet/foundation/hlp.foundation.rule-runtime/Evaluation/EvalSupport.cs
using System.Text.Json.Nodes;

using Harborline.Foundation.RuleEngine.Model;

namespace Harborline.Foundation.RuleEngine.Evaluation;

/// <summary>The resolution of a referenced cell handed to the evaluator. Public as the value
/// vocabulary an <see cref="Context.IContextAdapter"/> resolver speaks (ADR 0146 D3).</summary>
public readonly struct RefValue
{
    private RefValue(ValueState state, JsonNode? value, RuleError? error)
    {
        State = state;
        Value = value;
        Error = error;
    }

    public ValueState State { get; }
    public JsonNode? Value { get; }
    public RuleError? Error { get; }

    public static RefValue Resolved(JsonNode? value) => new(ValueState.Resolved, value, null);
    public static RefValue OfError(RuleError error) => new(ValueState.Error, null, error);
    public static RefValue Pending { get; } = new(ValueState.Pending, null, null);

    /// <summary>The ONE bad-aggregate refusal shape in this tier (ticket 162): an aggregate the
    /// evaluation context cannot provide refuses with <c>rule.bad_reference</c> carrying the
    /// address params <c>{agg: "section/fn/col"}</c>. Both resolvers (the form graph's
    /// CellResolver and the guard evaluator's ContextBagResolver) construct it HERE, so the
    /// param order cannot drift within the tier; the shared corpus pins the shape across tiers.</summary>
    public static RefValue UnavailableAggregate(string fn, string section, string col)
        => OfError(RuleError.Of(RuleEngineCodes.BadReference, "agg", section + "/" + fn + "/" + col));
}

/// <summary>
/// Resolves <c>var</c> / <c>agg</c> references for one rule evaluation against the
/// already-evaluated upstream cells + raw instance values (SPINE-1 design §2.2). Public as
/// the resolver an <see cref="Context.IContextAdapter"/> produces for its pillar data (ADR 0146 D3).
/// </summary>
public interface IValueResolver
{
    /// <summary>Resolves a canonical var path (<c>field.x</c> / <c>row.y</c> / <c>parent.f</c>).</summary>
    RefValue ResolveVar(string path);

    /// <summary>Resolves a table aggregate (<c>fn</c> over <c>section</c>.<c>col</c>).</summary>
    RefValue ResolveAgg(string fn, string section, string col);
}

/// <summary>
/// The shared step budget + wall-clock token for one whole-instance evaluation
/// (SPINE-1 design §4). One <see cref="EvalBudget"/> spans every cell + rule in the
/// instance so the 250k-op budget is a per-instance total, not per-rule.
/// </summary>
internal sealed class EvalBudget
{
    public EvalBudget(RuleEngineLimits limits, CancellationToken ct)
    {
        Limits = limits;
        CancellationToken = ct;
    }

    public RuleEngineLimits Limits { get; }
    public CancellationToken CancellationToken { get; }

    /// <summary>Steps consumed so far across the whole instance evaluation.</summary>
    public int Steps { get; private set; }

    /// <summary>Charges one step. The op-budget is the authoritative, outcome-affecting bound (fails
    /// closed with <see cref="RuleBudgetException"/>). The wall-clock / cancellation token is a
    /// non-authoritative liveness guard: it throws <see cref="RuleEngineTimeoutException"/>, which
    /// propagates as an infrastructure fault and NEVER becomes a divergent evaluation outcome
    /// (D1 ratification 2026-07-01).</summary>
    public void Charge()
    {
        if (CancellationToken.IsCancellationRequested) throw new RuleEngineTimeoutException();
        if (++Steps > Limits.StepBudget) throw new RuleBudgetException();
    }

    /// <summary>Charges <paramref name="size"/> steps proportional to a large operand's size — string
    /// concat and money arithmetic do O(size) … super-linear work for ONE AST node, so a single big
    /// operand must be charged proportionally rather than the flat one step (finding F4). Fails closed
    /// identically to <see cref="Charge"/>; the per-step add is clamped so it cannot overflow.</summary>
    public void ChargeSize(int size)
    {
        if (CancellationToken.IsCancellationRequested) throw new RuleEngineTimeoutException();
        Steps += size > Limits.StepBudget ? Limits.StepBudget + 1 : (size < 1 ? 1 : size);
        if (Steps > Limits.StepBudget) throw new RuleBudgetException();
    }
}

/// <summary>Per-cell evaluation state: the shared <see cref="EvalBudget"/> + this cell's resolver + clock.</summary>
internal sealed class EvalContext
{
    public EvalContext(IValueResolver resolver, DateTimeOffset now, EvalBudget budget)
    {
        Resolver = resolver;
        Now = now;
        Budget = budget;
    }

    public IValueResolver Resolver { get; }
    public DateTimeOffset Now { get; }
    public EvalBudget Budget { get; }

    /// <summary>Charges one step against the shared budget; fails closed on overflow/cancellation.</summary>
    public void Charge() => Budget.Charge();
}

/// <summary>A recoverable evaluation error that becomes a <c>ComputedValue.Error</c> / <c>Validity</c> failure.</summary>
internal sealed class RuleEvalException : Exception
{
    public RuleEvalException(RuleError error) => Error = error;
    public RuleError Error { get; }
}

/// <summary>Signals a referenced cell is <c>Pending</c> — the whole outcome becomes Pending.</summary>
internal sealed class RulePendingException : Exception;

/// <summary>Signals the per-instance step budget was exhausted — the authoritative, deterministic,
/// outcome-affecting bound (becomes a fail-closed <c>rule.budget_exceeded</c> result).</summary>
internal sealed class RuleBudgetException : Exception;
